ludwig/contribs/mlflow/__init__.py

```diff
@@ -161,15 +161,26 @@ def on_trainer_train_setup(self, trainer, save_path, is_coordinator):
     def on_eval_end(self, trainer, progress_tracker, save_path):
         if progress_tracker.steps not in self.logged_steps:
             self.logged_steps.add(progress_tracker.steps)
-            self.save_fn((progress_tracker.log_metrics(), progress_tracker.steps, save_path, True))  # Why True?
+            # Adds a tuple to the logging queue.
+            # True is passed to indicate that the background saving loop should continue.
+            self.save_fn((progress_tracker.log_metrics(), progress_tracker.steps, save_path, True))
 
     def on_trainer_train_teardown(self, trainer, progress_tracker, save_path, is_coordinator):
         if is_coordinator:
             if progress_tracker.steps not in self.logged_steps:
                 self.logged_steps.add(progress_tracker.steps)
-                self.save_fn((progress_tracker.log_metrics(), progress_tracker.steps, save_path, False))  # Why False?
-                if self.save_thread is not None:
-                    self.save_thread.join()
+                # Adds a tuple to the logging queue.
+                # False is passed to indicate that the background saving loop should break.
+                self.save_fn((progress_tracker.log_metrics(), progress_tracker.steps, save_path, False))
+            # False ensures that the background saving loop breaks.
+            # TODO(Justin): This should probably live in on_ludwig_end, once that's implemented.
+            self.save_fn((None, None, None, False))
+
+            # Close the save_thread.
+            if self.save_thread is not None:
+                self.save_thread.join(timeout=5)
+                if self.save_thread.is_alive():
+                    logger.warning("MLFlow save thread timed out and did not close properly.")
```

Comment:
This still leaves us with a potential hang. Should we modify this to be a [daemon thread](https://docs.python.org/3/library/threading.html#threading.Thread.daemon)? The downside there is that we wouldn't have detected this issue as quickly if the thread was just shut down on main thread exit.

Reply:
Since this seems to have been largely working with regular thread joining, I think we can defer the daemon thread change. If we start seeing additional warnings in our logs, happy to revisit though.